src/lib/sys.c

```diff
@@ -613,3 +614,60 @@ int sys_dir_is_empty(const char *path)
 	closedir(dir);
 	return ret;
 }
+
+int sys_write_file(char *file, void *content, size_t content_size)
+{
+	FILE *fp = NULL;
+	int ret = 0;
+
+	fp = fopen(file, "w");
+	if (!fp) {
+		return -errno;
+	}
+
+	if (fwrite(content, content_size, 1, fp) <= 0) {
+		ret = -1;
+		error("There was an error writing to file %s\n", file);
+	}
+
+	fclose(fp);
+	return ret;
+}
+
+void *sys_mmap_file(const char *file, size_t *file_length)
+{
+	struct stat st;
+	int fd = -1;
+	void *buffer = NULL;
+
+	fd = open(file, O_RDONLY);
+	if (fd == -1) {
+		debug("Failed to open %s: %s", file, strerror(errno));
+		goto error;
+	}
+
+	if (fstat(fd, &st) != 0) {
+		debug("Failed to stat %s file", file);
+		goto error;
+	}
+	*file_length = st.st_size;
+
+	buffer = mmap(NULL, *file_length, PROT_READ, MAP_PRIVATE, fd, 0);
+	if (buffer == MAP_FAILED) {
+		debug("Failed to mmap %s content", file);
```

Comment:
In this case your function will return MAP_FAILED and !template will be false, in the code bellow. I would change this function to return NULL on errors.

Reply:
ahh, I missed that, true!